django/db/migrations/state.py

```diff
@@ -563,15 +563,23 @@ def render(self, apps):
             )
         except LookupError:
             raise InvalidBasesError("Cannot resolve one or more bases from %r" % (self.bases,))
+
         # Clone fields for the body, add other bits.
         body = {name: field.clone() for name, field in self.fields.items()}
         body['Meta'] = meta
         body['__module__'] = "__fake__"
 
+        # resolve metaclass conflicts
+        base_types = {type(base) for base in bases}
+        metaclass_bases = tuple((
+            base for base in base_types if all((b == base or not issubclass(b, base) for b in base_types))
```

Comment:
I think this logic is flawed as `base_types` doesn't define an ordering so `metaclass_bases` is unstable and its order can make a difference in the mro of the metaclass.

Reply:
@charettes The reason I moved away from that is because of the way that bases of the model that are abstract aren't included in `self.bases`. Because of that, if the custom metaclass does some code depending on the abstract model then it will through an error (since it might not have properties the metaclass is expecting, but just during the migration). This way seems simpler, and since it's just to get the migration working, also seemed fine. The custom metaclass would still be necessary for models, but other parts of django will complain about that, not just the migration

Reply:
>  The reason I moved away from that is because of the way that bases of the model that are abstract aren't included in self.bases.

I think this is a different problem and that it should be tackled independently. If Django suppresses abstract bases when creating a `ModelState` it could still detect any metaclass that differs from `ModelBase` and pass it along as `metaclass`.

> This way seems simpler, and since it's just to get the migration working, also seemed fine.

I think the logic above is flawed though. The order of bases matter; you can't mix bases type to hack around the `TypeError` that would normally be raised on conflict.

Reply:
also I fixed the MRO of the generated metaclass so that should work

Reply:
@charettes so the problem is something like this:
```
class Abstract:
    @classmethod
    def a(cls):
        return 'a'
    class Meta:
        abstract = True

class Base(type):
    def __init__(cls, name, bases, attrs):
        attrs['ab'] = f'{cls.a()}b'
        super().__init__(name, bases, attrs)
```
so assuming our model is a descendant of `Abstract` and has as its metaclass `Base` then when it's recreated by `ModelState.from_model()` it will no longer has `Abstract` as a base. So it won't have the classmethod `a()` anymore either, so the `metaclass()` call in the `render()` function will fail